compiler/ir/stream/access_pattern.py

```diff
@@ -0,0 +1,242 @@
+        For `dim` = 1, `template_bound` = 2:
+            [2, 4, 2, 2]
+
+        """
+        transform_map = AffineMap(
+            num_dims=self.num_dims + 1,
+            num_symbols=0,
+            # (d0, d1, d2, ..., dim-1) -> (d0, d1, d2, ..., dim-1)
+            results=tuple(AffineDimExpr(i) for i in range(dim))
+            # (dim) -> (template_bound * dim + dim + 1)
+            + (AffineDimExpr(dim) * template_bound + AffineDimExpr(dim + 1),)
+            # (dim + 1, dim + 2, ...) -> (dim + 2, dim + 3, dim + 3)
+            + tuple(AffineDimExpr(i + 1) for i in range(dim + 1, self.num_dims)),
+        )
+        new_pattern = self.pattern.compose(transform_map)
+        bound_to_tile = self.bounds[dim]
+        tiled_bound = bound_to_tile // template_bound if bound_to_tile else None
+        new_bounds = (
+            self.bounds[:dim] + (tiled_bound, template_bound) + self.bounds[dim + 1 :]
+        )
+
+        return type(self)(new_bounds, new_pattern)
+
+
+@dataclass(frozen=True)
+class SchedulePattern(AccessPattern):
+    """
+    A schedule pattern is a pattern for a schedule of an operation.
+
+    Schedule patterns are constrained to have static bounds.
+    """
+
+    # constrain bounds to only be int
+    bounds: tuple[int, ...]
+
+    def __init__(self, bounds: Sequence[int], pattern: AffineMap):
+        if any(bound is None or bound <= 0 for bound in bounds):
+            raise ValueError(
+                "All bounds must be static, strictly positive integers for a schedule"
+            )
+
+        super().__init__(bounds, pattern)
+
+
+@dataclass(frozen=True)
+class TemplatePattern(AccessPattern):
+    """
+    Template pattern is a pattern for an accelerator template.
+
+    Templates should not be transformed through either tiling/rotating/others.
+    """
+
+    def __init__(self, bounds: Sequence[int], pattern: AffineMap):
+        super().__init__(bounds, pattern)
+
+    def tile_dim(self, dim: int, template_bound: int) -> Self:
+        raise RuntimeError("A template should not be tiled")
+
+    def rotate(self, dim: int) -> Self:
+        raise RuntimeError("A template should not be rotated")
+
+    def matches(self, sp: SchedulePattern):
+        """
+        Check if a given schedule pattern matches this
+        template pattern.
+        """
+        if sp.num_dims != self.num_dims:
+            return False
+        if sp.pattern != self.pattern:
+            return False
+        return True
+
+
+P = TypeVar("P", bound=AccessPattern)
+
+
+class PatternCollection(Sequence[P], Generic[P], ABC):
+    """
+    Abstract base class for collections of AccessPatterns.
+    Provides common methods and properties for Schedule and Template classes.
+    """
+
+    def __init__(self, patterns: Iterable[P]):
+        self._patterns = tuple(patterns)
+
+    @overload
+    def __getitem__(self, index: int) -> P:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> tuple[P]:
+        ...
```

Comment:
oof, is this how you're supposed to tell pyright? Cool! but oof!?

Reply:
yes: https://docs.python.org/3/library/typing.html#typing.overload
but i find this quite nice? how else would you specify specific different return types for different input types